internal/wrphandlers/qos/qos.go

```diff
@@ -0,0 +1,190 @@
+// SPDX-FileCopyrightText: 2024 Comcast Cable Communications Management, LLC
+// SPDX-License-Identifier: Apache-2.0
+
+package qos
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"github.com/xmidt-org/wrp-go/v3"
+	"github.com/xmidt-org/xmidt-agent/internal/websocket"
+	"github.com/xmidt-org/xmidt-agent/internal/wrpkit"
+)
+
+var (
+	ErrInvalidInput     = errors.New("invalid input")
+	ErrMisconfiguredQOS = errors.New("misconfigured QOS")
+	ErrQOSHasShutdown   = errors.New("QOS has been shutdown")
+)
+
+// Option is a functional option type for QOS.
+type Option interface {
+	apply(*Handler) error
+}
+
+type optionFunc func(*Handler) error
+
+func (f optionFunc) apply(c *Handler) error {
+	return f(c)
+}
+
+type serviceQOSHandler func(wrp.Message) (<-chan wrp.Message, <-chan struct{})
+
+// Handler queues incoming messages and sends them to the next wrphandler
+type Handler struct {
+	next serviceQOSHandler
+	// queue for wrp messages, ingested by serviceQOS
+	queue chan wrp.Message
+	// maxQueueSize is the allowable max size of the qos' priority queue, based on the sum of all queued wrp message's payload
+	maxQueueSize int
+
+	lock   sync.Mutex
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// New creates a new instance of the Handler struct.  The parameter next is the
+// handler that will be called and monitored for errors.
+// Note, once cancel is called, any calls to Handler.HandleWrp will result in
+// an ErrQOSHasShutdown error
+func New(next websocket.Egress, opts ...Option) (h *Handler, err error) {
+	if next == nil {
+		return nil, ErrInvalidInput
+	}
+
+	h = &Handler{
+		next: handleWRPWrapper(next),
+	}
+
+	var errs error
+	for _, opt := range opts {
+		if opt != nil {
+			if err := opt.apply(h); err != nil {
+				errs = errors.Join(errs, err)
+				h = nil
+			}
+		}
+	}
+
+	return h, errs
+}
+
+func (h *Handler) Start() error {
+	h.lock.Lock()
+	defer h.lock.Unlock()
+
+	if h.ctx != nil {
+		return nil
+	}
+
+	// h.queue will never block as long as the serviceQOS goroutine is running.
+	h.queue = make(chan wrp.Message, 1)
+	// h.cancel() stops serviceQOS by closing its `done` chan.
+	h.ctx, h.cancel = context.WithCancel(context.Background())
+	go serviceQOS(h.ctx.Done(), h.queue, h.maxQueueSize, h.next)
```

Comment:
I'd change `serviceQOS` to a method like `func (h *Handler) run()`.  It's simpler & what folks expect to see.